src/osdc/ObjectCacher.cc

```diff
@@ -640,12 +649,19 @@ void ObjectCacher::Object::discard(loff_t off, loff_t len,
       // we should mark all Rx bh to zero
       continue;
     } else {
-      ceph_assert(bh->waitfor_read.empty());
+      for ([[maybe_unused]] auto& [off, ctxs] : bh->waitfor_read) {
```

Comment:
Why are the reads still pending at this point? Shouldn't they have failed? To me, it seems like a bug in assumptions if you are attempting to concurrently discard a section that you have a pending IO against.

Reply:
FWIW, in `librbd` we explicitly handle this discard case via a call to `discard_writeback` [1] prior to invoking the `discard_set` operation since we had similar races before (on test cases w/ random IO causing overlapping read/writes/discards). 

`librbd` doesn't use truncate so I have no skin in that fight. 

[1] https://github.com/ceph/ceph/blob/master/src/librbd/cache/ObjectCacherObjectDispatch.cc#L288

Reply:
I may be misunderstanding but in one stacktrace the client actually **is** calling `discard_writeback` and that generates the assertion. See the summary of the ticket:

https://tracker.ceph.com/issues/46434

So I'm not sure how we'd change the code in `Client` to get around this assertion?

Reply:
Ah, in that case disregard.